lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/AvroFilePageIterator.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.hive.formats.avro;
+
+import io.trino.spi.Page;
+import org.apache.avro.AvroRuntimeException;
+import org.apache.avro.Schema;
+import org.apache.avro.file.DataFileReader;
+import org.apache.avro.file.SeekableInput;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+public class AvroFilePageIterator
+{
+    private final AvroPageDataReader dataReader;
+    private final DataFileReader<Optional<Page>> fileReader;
+    private Optional<Page> nextPage = Optional.empty();
+    private long stop;
+
+    public AvroFilePageIterator(Schema readerSchema, AvroTypeManager typeManager, SeekableInput inputStream)
+            throws IOException, AvroTypeException
+    {
+        this(readerSchema, typeManager, 0, -1, inputStream);
+    }
+
+    // Calling class responsible for closing input stream
+    public AvroFilePageIterator(Schema readerSchema, AvroTypeManager typeManager, long start, long stop, SeekableInput inputStream) throws IOException, AvroTypeException
+    {
+        this.stop = stop;
+        dataReader = new AvroPageDataReader(readerSchema, typeManager);
+        try {
+            fileReader = new DataFileReader<>(inputStream, dataReader);
+            fileReader.sync(start);
+        }
+        catch (AvroPageDataReader.AvroTypeExceptionRuntimeWrapper runtimeWrapper) {
+            throw runtimeWrapper.getAvroTypeException();
```

Comment:
Add a comment like `Avro Dataum interface can't throw checked exceptions, so the exception is wrapped in a runtime exception that must be unwrapped`